### PR TITLE
Stop natural spawning after limit has been reached

### PR DESCRIPTION
**Pull request by kickash32**
First patch PR for me... so not sure if I did this right.
This patch changes natural mob spawning so that it only spawns up to the the limit specified per world without any overshooting. This prevents the notorious vanilla bug where in 300+ fish are spawned in an ocean while the limit is supposed to be set to 15.

If this patch is accepted, we should look into advising players to increase the watermob limit so they are not surprised by lowered spawn rates.

It accomplishes this by tracking the number of mobs spawned by each chunk, and stopping spawns once the world's mob limit is reached (for that type of mob).

This is a resubmission of #2119

### CODE REVIEW

**Comment by kickash32** [comment]
`Applying: Fix Mob Spawning Relative to View Distance
error: sha1 information is lacking or useless (src/main/java/net/minecraft/server/ChunkProviderServer.java).
error: could not build fake ancestor
Patch failed at 0010 Fix Mob Spawning Relative to View Distance`

Really not sure why this check failed and only on the 2nd commit...

**Comment by GMatrixGames** [comment]
It failed because you modified the patch directly through the github editor.

**Comment by electronicboy** [comment]
Spigot made some mapping changes, I've flushed the cache so it should build this time

File: Spigot-Server-Patches/0398-Actually-Limit-Natural-Spawns-To-Limit.patch [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+                                 if (enumcreaturetype != EnumCreatureType.MISC && (!enumcreaturetype.c() || this.allowAnimals) && (enumcreaturetype.c() || this.allowMonsters) && (!enumcreaturetype.d() || flag2)) {
+                                     int k1 = limit * l / ChunkProviderServer.b; // CraftBukkit - use per-world limits
+ 
+-                                    if (object2intmap.getInt(enumcreaturetype) <= k1) {
+-                                        SpawnerCreature.a(enumcreaturetype, (World) this.world, chunk, blockposition);
++                                    // Paper start - only allow spawns upto the limit per chunk and update count afterwards
++                                    int difference = k1 - object2intmap.getInt(enumcreaturetype);
++                                    if (difference > 0) {
++                                        int tmp = SpawnerCreature.a(enumcreaturetype, (World) this.world, chunk, blockposition, difference + 1);
++                                        object2intmap.put(enumcreaturetype, object2intmap.getInt(enumcreaturetype) + tmp);
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0398-Actually-Limit-Natural-Spawns-To-Limit.patch]: cache the value returned by the map, there's no risk of it changing in the call above
alternatively look into using the compute() method of the map

File: Spigot-Server-Patches/0398-Actually-Limit-Natural-Spawns-To-Limit.patch [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+@@ -360,8 +360,12 @@ public class ChunkProviderServer extends IChunkProvider {
+                                 if (enumcreaturetype != EnumCreatureType.MISC && (!enumcreaturetype.c() || this.allowAnimals) && (enumcreaturetype.c() || this.allowMonsters) && (!enumcreaturetype.d() || flag2)) {
+                                     int k1 = limit * l / ChunkProviderServer.b; // CraftBukkit - use per-world limits
+ 
+-                                    if (object2intmap.getInt(enumcreaturetype) <= k1) {
+-                                        SpawnerCreature.a(enumcreaturetype, (World) this.world, chunk, blockposition);
++                                    // Paper start - only allow spawns upto the limit per chunk and update count afterwards
++                                    int difference = k1 - object2intmap.getInt(enumcreaturetype);
++                                    if (difference > 0) {
++                                        int tmp = SpawnerCreature.a(enumcreaturetype, (World) this.world, chunk, blockposition, difference + 1);
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0398-Actually-Limit-Natural-Spawns-To-Limit.patch]: should just pass the difference and not `difference +1`

File: Spigot-Server-Patches/0398-Actually-Limit-Natural-Spawns-To-Limit.patch (hunk 1 above)
Comment (kickash32) [reply Spigot-Server-Patches/0398-Actually-Limit-Natural-Spawns-To-Limit.patch]: object2intmap is itself a map of cached values of entity types to counts, I don't see how avoiding updating it achieves anything.

File: Spigot-Server-Patches/0398-Actually-Limit-Natural-Spawns-To-Limit.patch (hunk 2 above)
Comment (kickash32) [reply Spigot-Server-Patches/0398-Actually-Limit-Natural-Spawns-To-Limit.patch]: Done

File: Spigot-Server-Patches/0398-Actually-Limit-Natural-Spawns-To-Limit.patch (hunk 1 above)
Comment (Spottedleaf) [reply Spigot-Server-Patches/0398-Actually-Limit-Natural-Spawns-To-Limit.patch]: I never made such an argument

Comment (kickash32) [reply Spigot-Server-Patches/0398-Actually-Limit-Natural-Spawns-To-Limit.patch]: Oh sorry I misunderstood what you meant, I'll make the changes.

Comment (kickash32) [reply Spigot-Server-Patches/0398-Actually-Limit-Natural-Spawns-To-Limit.patch]: Done

**Comment by kickash32** [comment]
refactored to make ABI changes compatible

**Review by electronicboy — APPROVED** [review]

**Review by zachbr — APPROVED** [review]

**Review by Spottedleaf — APPROVED** [review]

**Comment by Sikatsu** [comment]
So do we need to change any settings like the watermob limit now that this is merged?

**Comment by electronicboy** [comment]
`[9:03 AM] electroniccat: if you reduced the amount to workaround the bugs that existed without that, then yea, you probably do`

**Comment by Sikatsu** [comment]
Yeah I set that back to default at the moment after you replied on discord, but I wondered if it needs to be increased above the default due to this. Just to make sure ^^

**Comment by aeonlamb** [comment]
Is this really considered a "bug" with vanilla? If I understand the nature of this patch correctly, it'll halt a packspawn attempt if that packspawn would go over the mob cap. Or perhaps just stop the packspawn attempt halfway if the next mob would go over the limit or something.

The reason that more mobs can spawn than the mob cap limit is due to how the game can spawn multiple entities and do multiple spawn attempts in the same tick, I don't think this is a "bug". Maybe it seems wrong or incorrect to some people but I think mob spawning has been that way for a very, very long time. If it were a vanilla bug they would've fixed it. I feel like if something like this is included in paper, it should at least be made to be configurable, because it messes with vanilla mechanics quite a lot in a way some server owners (or players) may not like.

If i'm misunderstanding the purpose of this patch though, feel free to correct me.

**Comment by kickash32** [comment]
I'd definitely say its a bug, because if you sit in an ocean 300+ fish would spawn and then slowly despawn down to 15, then suddenly 300+ fish would spawn again and the cycle would repeat. I suspect the reason Mojang never fixed this is because in vanilla no one can easily count the mobs and the limits are only available on the wiki. Thus no one reported this issue and Mojang never looked into it (it was only really a major issue once they released 1.13).

**Comment by kickash32** [comment]
On a side note, it seems this vanilla bug https://bugs.mojang.com/browse/MC-151144 (fixed in 1.14.3 MC) is causing frustration to server owners. Not sure how important this bug is in relation to this patch, but this patch should still make mobspawning better than it was before even with this bug...

**Pull request merged**